contracts/Staking.sol

```diff
@@ -0,0 +1,191 @@
+        }
+    }
+
+    /*
+     * Removes the deposit of the user and sends the amount of `tokenAddress` back to the `user`
+     */
+    function withdraw(address tokenAddress) public {
+        require(balances[msg.sender][tokenAddress] > 0, "Staking: User has empty balance");
+
+        uint256 amount = balances[msg.sender][tokenAddress];
+        balances[msg.sender][tokenAddress] = 0;
+
+        IERC20 token = IERC20(tokenAddress);
+        token.transfer(msg.sender, amount);
+
+        // epoch logic
+        uint256 currentEpoch = getCurrentEpoch();
+
+        if (!epochIsInitialized(tokenAddress, currentEpoch)) {
+            address[] memory tokens = new address[](1);
+            tokens[0] = tokenAddress;
+            manualEpochInit(tokens, currentEpoch);
+        }
+
+        // decrease the balance this user contributed to the poolSize at the beginning of the current epoch == the epoch balance of previous epoch
+        poolSize[tokenAddress][currentEpoch].size -= getEpochUserBalance(msg.sender, tokenAddress, currentEpoch);
+
+        // update the pool size of the next epoch to its current balance
+        Pool storage pNextEpoch = poolSize[tokenAddress][currentEpoch + 1];
+        pNextEpoch.size = token.balanceOf(address(this));
+        pNextEpoch.set = true;
+
+        // remove any contribution for the user in the current epoch
+        Checkpoint[] storage checkpoints = balanceCheckpoints[msg.sender][tokenAddress];
+        uint256 last = checkpoints.length - 1;
+
+        // note: it's impossible to have a withdraw and no checkpoints because the balance would be 0 and revert
+
+        // there was a deposit in an older epoch (more than 1 behind [eg: previous 0, now 5]) but no other action since then
+        if (checkpoints[last].epochId < currentEpoch) {
+            checkpoints.push(Checkpoint(currentEpoch, 0));
+        }
+        // there was a deposit in the `epochId - 1` epoch => we have a checkpoint for the current epoch
+        else if (checkpoints[last].epochId == currentEpoch) {
+            checkpoints[last].balance = 0;
+        }
+        // there was a deposit in the current epoch
+        else {
+            checkpoints[last].balance = 0;
+        }
+    }
+
+    function getEpochUserBalance(address user, address token, uint256 epochId) public view returns (uint256) {
+        Checkpoint[] storage checkpoints = balanceCheckpoints[user][token];
+
+        // if there are no checkpoints, it means the user never deposited any tokens, so the balance is 0
+        if (checkpoints.length == 0 || epochId < checkpoints[0].epochId) {
+            return 0;
+        }
+
+        uint min = 0;
+        uint max = checkpoints.length - 1;
+
+        // shortcut for blocks newer than the latest checkpoint == current balance
+        if (epochId >= checkpoints[max].epochId) {
+            return checkpoints[max].balance;
+        }
+
+        // binary search of the value in the array
+        while (max > min) {
```

Comment:
This still makes me anxious, but given that it's per user and won't affect everyone, and the number of checkpoints for a user is limited, we should be ok.

Reply:
Yeah, it's per user **and per epoch** so we limit it quite a lot. It's going to be a long time before it ends up using a lot of gas.